File: 2-D DP/Matrix Chain/Burst Balloons/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int maxCoins(vector<int> &nums) {
    vector<int> paddedNums;
    paddedNums.push_back(1);
    for (int x : nums)
      paddedNums.push_back(x);
    paddedNums.push_back(1);

    int n = paddedNums.size();
    vector<vector<int>> dp(n, vector<int>(n, 0));

    for (int len = 1; len <= nums.size(); ++len) {
      for (int left = 0; left < n - len - 1; ++left) {
        int right = left + len + 1;
        for (int k = left + 1; k < right; ++k) {
          dp[left][right] =
              max(dp[left][right],
                  paddedNums[left] * paddedNums[k] * paddedNums[right] +
                      dp[left][k] + dp[k][right]);
        }
      }
    }

    return dp[0][n - 1];
  }
};

int main() {
  Solution sol;

  vector<int> nums1 = {3, 1, 5, 8};
  cout << "Test Case 1: " << sol.maxCoins(nums1) << endl; // Expect 167

  vector<int> nums2 = {1, 5};
  cout << "Test Case 2: " << sol.maxCoins(nums2) << endl; // Expect 10

  return 0;
}

File: 2-D DP/Matrix Chain/Burst Balloons/solution.java
class Solution {
    public int maxCoins(int[] nums) {
        int n = nums.length;
        int[] paddedNums = new int[n + 2];
        paddedNums[0] = 1;
        paddedNums[n + 1] = 1;
        for (int i = 0; i < n; i++) paddedNums[i + 1] = nums[i];
        
        int[][] dp = new int[n + 2][n + 2];
        
        for (int len = 1; len <= n; len++) {
            for (int left = 0; left < n + 2 - len - 1; left++) {
                int right = left + len + 1;
                for (int k = left + 1; k < right; k++) {
                    dp[left][right] = Math.max(dp[left][right], 
                        paddedNums[left] * paddedNums[k] * paddedNums[right] + dp[left][k] + dp[k][right]);
                }
            }
        }
        
        return dp[0][n + 1];
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] nums1 = {3,1,5,8};
        System.out.println("Test Case 1: " + sol.maxCoins(nums1)); // 167
        int[] nums2 = {1,5};
        System.out.println("Test Case 2: " + sol.maxCoins(nums2)); // 10
    }
}
